### PR TITLE
Change type of `Window` `background` from `color` to `brush`

### LINKED ISSUE #1681: Window background property should have the same type as Rectangle's background property
Rectangle's background property is a brush, while Window's background property is only a color. For consistency, the latter should have the same type as the former. That means that we have to write extra code to render the background "by hand" if it's not a solid color for use with Qt or glClear().

### PR DESCRIPTION
**Pull request by FloVanGH**
The background type of `Window` is now `background`. It's not a breaking change. At the moment  the software renderer does not provide gradient drawing and until it is implemented a gradient window background will be drawn as solid color. Fixes #1681.

### CODE REVIEW

File: internal/backends/winit/renderer/femtovg.rs [hunk 1]
```diff
@@ -143,32 +144,41 @@ impl super::WinitCompatibleRenderer for FemtoVGRenderer {
         let window = WindowInner::from_pub(window_adapter.window());
 
         window.draw_contents(|components| {
+            let is_solid_background = window
+                .window_item()
+                .map_or_else(|| false, |w| w.as_pin_ref().background().is_solid_color());
+
```
Comment (tronical) [inline internal/backends/winit/renderer/femtovg.rs:150]: Every time we query the background() property, a new property dependency is added. I suggest to address that and also simplify the code a little by doing the `as_pin_ref().background()` done only once:

```rust
let window_background_brush = window.window_item().map(|w| w.as_pin_ref().background());
```

And then later for example:

```rust
if let Some(Brush::SolidColor(clear_color)) = window_background_brush {
    femtovg_canvas.clear_rect(...);
}
```

as well as

```rust
match window_background_color {
    Some(Brush::SolidColor(..)) | None => {},
    brush @ _ => {
        draw_rect(..., brush);
    }
}
```

**Review by tronical — COMMENTED** [review]
Looks like the right direction to me. This also needs an implementation for Skia and Qt. Let me know if you'd like me to take care of one of these.

File: internal/backends/winit/renderer/femtovg.rs (hunk 1 above)
Comment (FloVanGH) [reply internal/backends/winit/renderer/femtovg.rs]: done

**Comment by ogoffart** [comment]
> It's not a breaking change.

Unless this was used in an alias:
```
X := Window { 
   property <color> foo <=> background;
}
```

Is that Ok?

File: internal/core/software_renderer.rs [hunk 2]
```diff
@@ -318,7 +319,8 @@ fn render_window_frame_by_line<const MAX_BUFFER_AGE: usize>(
             dirty_region.min_x() as usize..dirty_region.max_x() as usize,
             |line_buffer| {
                 let offset = dirty_region.min_x() as usize;
-                TargetPixel::blend_slice(line_buffer, background.into());
+                // FIXME gradient
+                TargetPixel::blend_slice(line_buffer, background.color().into());
```
Comment (ogoffart) [inline internal/core/software_renderer.rs:323]: Micro optimization. Would be nice to not call background.color().into()  on every line.  In fact, the PremultipliedRgbaColor color could be cached outside of the `while`   (but maybe the compiler is able to do that anyway?)

File: internal/core/graphics/brush.rs [hunk 3]
```diff
@@ -53,6 +53,19 @@ impl Brush {
...
+    ///
+    /// ```
+    /// # use i_slint_core::graphics::*;
+    /// assert!(Brush::SolidColor(Color::from_argb_u8(0, 255, 128, 140)).is_solid_color());
+    /// ```
+    pub fn is_solid_color(&self) -> bool {
+        match self {
+            Brush::SolidColor(_) => true,
+            _ => false,
+        }
```
Comment (ogoffart) [inline internal/core/graphics/brush.rs:66]:
```suggestion
        matches!(self, Brush::SolidColor(_))
```


This is exposed in the public API, not sure if worth adding this function.
It also doesn't seem to be used.

**Comment by FloVanGH** [comment]
@tronical for skia it's now also implemented. I will check qt next. Thank you.

File: internal/backends/qt/qt_window.rs [hunk 4]
```diff
@@ -1532,7 +1532,18 @@ impl WindowAdapterSealed for QtWindow {
                 size.height = existing_size.height;
             }
         }
-        let background: u32 = window_item.background().as_argb_encoded();
+
+        let window_background_brush =
+            window_inner.window_item().map(|w| w.as_pin_ref().background());
```
Comment (ogoffart) [inline internal/backends/qt/qt_window.rs:1537]: we already have window_item in a local variable, so no need to call map and stuff.
Also with Qt we can just transform the background brush in a QBrush (there should already have a helper function for that) and set that as a palette for the window background

**Comment by FloVanGH** [comment]
> > It's not a breaking change.
> 
> Unless this was used in an alias:
> 
> ```
> X := Window { 
>    property <color> foo <=> background;
> }
> ```
> 
> Is that Ok?

@tronical what do you think? I'm not sure.

File: internal/backends/qt/qt_window.rs (hunk 4 above)
Comment (FloVanGH) [reply internal/backends/qt/qt_window.rs]: done thank you

File: internal/core/graphics/brush.rs (hunk 3 above)
Comment (FloVanGH) [reply internal/core/graphics/brush.rs]: done

File: internal/core/software_renderer.rs (hunk 2 above)
Comment (FloVanGH) [reply internal/core/software_renderer.rs]: I have done this optimization.

File: internal/core/graphics/brush.rs [hunk 5]
```diff
@@ -53,6 +53,16 @@ impl Brush {
         }
     }
 
+    /// Returns true if this brush is a solid color
+    ///
+    /// ```
+    /// # use i_slint_core::graphics::*;
+    /// assert!(Brush::SolidColor(Color::from_argb_u8(0, 255, 128, 140)).is_solid_color());
+    /// ```
+    pub fn is_solid_color(&self) -> bool {
```
Comment (ogoffart) [inline internal/core/graphics/brush.rs:62]: Why do we need this function?

Comment (FloVanGH) [reply internal/core/graphics/brush.rs:62]: It is not yet used. It's not problem to remove it.

**Comment by ogoffart** [comment]
Looks good.  
Deserve a changelog entry.

**Review by ogoffart — APPROVED** [review]
(please squash when merging)

**Pull request merged**